### ISSUE TITLE
rst2pdf crashes with several documents from the Linux Kernel

### ISSUE THREAD

**Issue by mchehab** [rst2pdf/rst2pdf#958]
Trying to build pdf docs from Linux .rst files doesn't work.

**Description of problem**

Cloning this tree, which contains a patch adding rst2pdf support:

   https://git.linuxtv.org/mchehab/experimental.git/log/?h=rst2pdf-v3

and building the documentation with:
```
   $ git clone git://linuxtv.org/mchehab/experimental.git -b rst2pdf-v3
   $ make rst2pdf
```
Should create 71 pdf files under `Documentation/output/pdf` directory. 18 of them have zero size, because rst2pdf crashed.

This is an example of one of such crashes.

```
[ERROR] pdfbuilder.py:159 Failed to build doc
Traceback (most recent call last):
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/rst2pdf/pdfbuilder.py", line 156, in write
    docwriter.write(doctree, destination)
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/docutils/writers/__init__.py", line 78, in write
    self.translate()
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/rst2pdf/pdfbuilder.py", line 672, in translate
    createpdf.RstToPdf(
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/rst2pdf/createpdf.py", line 659, in createPdf
    pdfdoc.multiBuild(elements)
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/reportlab/platypus/doctemplate.py", line 1177, in multiBuild
    self.canv.save()
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/reportlab/pdfgen/canvas.py", line 1277, in save
    self._doc.SaveToFile(self._filename, self)
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/reportlab/pdfbase/pdfdoc.py", line 219, in SaveToFile
    data = self.GetPDFData(canvas)
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/reportlab/pdfbase/pdfdoc.py", line 245, in GetPDFData
    return self.format()
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/reportlab/pdfbase/pdfdoc.py", line 427, in format
    IOf = IO.format(self)
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/reportlab/pdfbase/pdfdoc.py", line 876, in format
    fcontent = format(self.content, document, toplevel=1)   # yes this is at top level
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/reportlab/pdfbase/pdfdoc.py", line 80, in format
    f = element.format(document)
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/reportlab/pdfbase/pdfdoc.py", line 1613, in format
    return D.format(document)
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/reportlab/pdfbase/pdfdoc.py", line 683, in format
    L = [(format(PDFName(k),document)+b" "+format(dict[k],document)) for k in keys]
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/reportlab/pdfbase/pdfdoc.py", line 683, in <listcomp>
    L = [(format(PDFName(k),document)+b" "+format(dict[k],document)) for k in keys]
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/reportlab/pdfbase/pdfdoc.py", line 80, in format
    f = element.format(document)
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/reportlab/pdfbase/pdfdoc.py", line 1778, in format
    if f is None: raise ValueError("format not resolved, probably missing URL scheme or undefined destination target for '%s'" % self.name)
ValueError: format not resolved, probably missing URL scheme or undefined destination target for 'sleeping-things'
```

🖥 **Versions**

```
$ python -V
Python 3.9.0
$ pip freeze | grep -E '(rst2pdf|reportlab|[Ss]phinx)'
reportlab==3.5.56
rst2pdf==0.98
Sphinx==2.4.4
sphinx-rtd-theme==0.5.0
sphinxcontrib-applehelp==1.0.2
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==1.0.3
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml==1.1.4
```

Tested on Fedora 33.

**Comment by mchehab** [comment]
The output of the build is shown below:

```
$ ls -l Documentation/output/pdf/
total 26120
-rw-rw-r-- 1 mchehab mchehab   82714 dez 10 17:12 accounting.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:09 admin-guide.pdf
-rw-rw-r-- 1 mchehab mchehab  564458 dez 10 17:08 arm64.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:10 arm.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:10 block.pdf
-rw-rw-r-- 1 mchehab mchehab  536263 dez 10 17:07 bpf.pdf
-rw-rw-r-- 1 mchehab mchehab  208210 dez 10 17:09 cdrom.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:07 core-api.pdf
-rw-rw-r-- 1 mchehab mchehab   78644 dez 10 17:09 cpu-freq.pdf
-rw-rw-r-- 1 mchehab mchehab  759440 dez 10 17:12 crypto.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:07 devicetree.pdf
-rw-rw-r-- 1 mchehab mchehab 1113261 dez 10 17:09 dev-tools.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:10 doc-guide.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:10 driver-api.pdf
-rw-rw-r-- 1 mchehab mchehab   98318 dez 10 17:10 fault-injection.pdf
-rw-rw-r-- 1 mchehab mchehab  782364 dez 10 17:10 fb.pdf
-rw-rw-r-- 1 mchehab mchehab 5208004 dez 10 17:11 filesystems.pdf
-rw-rw-r-- 1 mchehab mchehab  358350 dez 10 17:09 firmware-guide.pdf
-rw-rw-r-- 1 mchehab mchehab   68519 dez 10 17:07 fpga.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:11 gpu.pdf
-rw-rw-r-- 1 mchehab mchehab  298264 dez 10 17:08 hid.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:08 hwmon.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:09 i2c.pdf
-rw-rw-r-- 1 mchehab mchehab  235541 dez 10 17:09 ia64.pdf
-rw-rw-r-- 1 mchehab mchehab  105156 dez 10 17:08 ide.pdf
-rw-rw-r-- 1 mchehab mchehab   25279 dez 10 17:08 iio.pdf
-rw-rw-r-- 1 mchehab mchehab  103740 dez 10 17:08 infiniband.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:09 input.pdf
-rw-rw-r-- 1 mchehab mchehab   67318 dez 10 17:07 isdn.pdf
-rw-rw-r-- 1 mchehab mchehab  562043 dez 10 17:10 kbuild.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:09 kernel-hacking.pdf
-rw-rw-r-- 1 mchehab mchehab  237262 dez 10 17:10 leds.pdf
-rw-rw-r-- 1 mchehab mchehab  183535 dez 10 17:11 livepatch.pdf
-rw-rw-r-- 1 mchehab mchehab  642626 dez 10 17:12 locking.pdf
-rw-rw-r-- 1 mchehab mchehab  159492 dez 10 17:07 m68k.pdf
-rw-rw-r-- 1 mchehab mchehab  100760 dez 10 17:08 maintainer.pdf
-rw-rw-r-- 1 mchehab mchehab   55201 dez 10 17:09 mhi.pdf
-rw-rw-r-- 1 mchehab mchehab   21269 dez 10 17:10 mips.pdf
-rw-rw-r-- 1 mchehab mchehab  345754 dez 10 17:07 misc-devices.pdf
-rw-rw-r-- 1 mchehab mchehab   79128 dez 10 17:12 netlabel.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:11 networking.pdf
-rw-rw-r-- 1 mchehab mchehab   26944 dez 10 17:10 openrisc.pdf
-rw-rw-r-- 1 mchehab mchehab   42861 dez 10 17:08 parisc.pdf
-rw-rw-r-- 1 mchehab mchehab  458735 dez 10 17:11 PCI.pdf
-rw-rw-r-- 1 mchehab mchehab   56467 dez 10 17:08 pcmcia.pdf
-rw-rw-r-- 1 mchehab mchehab  896045 dez 10 17:08 powerpc.pdf
-rw-rw-r-- 1 mchehab mchehab  832148 dez 10 17:08 power.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:09 process.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:09 RCU.pdf
-rw-rw-r-- 1 mchehab mchehab   56573 dez 10 17:12 riscv.pdf
-rw-rw-r-- 1 mchehab mchehab  576367 dez 10 17:08 s390.pdf
-rw-rw-r-- 1 mchehab mchehab  426567 dez 10 17:09 scheduler.pdf
-rw-rw-r-- 1 mchehab mchehab 1898403 dez 10 17:08 scsi.pdf
-rw-rw-r-- 1 mchehab mchehab  554974 dez 10 17:09 security.pdf
-rw-rw-r-- 1 mchehab mchehab   88367 dez 10 17:07 sh.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:07 sound.pdf
-rw-rw-r-- 1 mchehab mchehab  205993 dez 10 17:12 sparc.pdf
-rw-rw-r-- 1 mchehab mchehab  167833 dez 10 17:08 spi.pdf
-rw-rw-r-- 1 mchehab mchehab  416038 dez 10 17:08 staging.pdf
-rw-rw-r-- 1 mchehab mchehab  326804 dez 10 17:07 target.pdf
-rw-rw-r-- 1 mchehab mchehab  133420 dez 10 17:07 timers.pdf
-rw-rw-r-- 1 mchehab mchehab 1642804 dez 10 17:07 trace.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:10 translations.pdf
-rw-rw-r-- 1 mchehab mchehab  756579 dez 10 17:10 usb.pdf
-rw-rw-r-- 1 mchehab mchehab       0 dez 10 17:07 userspace-api.pdf
-rw-rw-r-- 1 mchehab mchehab 1932509 dez 10 17:12 virt.pdf
-rw-rw-r-- 1 mchehab mchehab  677027 dez 10 17:10 vm.pdf
-rw-rw-r-- 1 mchehab mchehab  157999 dez 10 17:10 w1.pdf
-rw-rw-r-- 1 mchehab mchehab  236123 dez 10 17:09 watchdog.pdf
-rw-rw-r-- 1 mchehab mchehab  953345 dez 10 17:07 x86.pdf
-rw-rw-r-- 1 mchehab mchehab   46073 dez 10 17:12 xtensa.pdf
```

**Comment by lornajane** [comment]
Thanks for the detailed bug report. Do all the failing files fail with the same error message? I'm not sure we've tested against 3.9 yet but we definitely should do so - and I'm not sure the version can be blamed here in any case.

**Comment by mchehab** [comment]
> Thanks for the detailed bug report. Do all the failing files fail with the same error message?

Yes. They're all at line 156 of pdfbuilder.py. I'm enclosing the full log from rst2pdf.

> I'm not sure we've tested against 3.9 yet but we definitely should do so - and I'm not sure the version can be blamed here in any case.

I suspect that this is not related to Python 3.9. I tested `rst2pdf` in the past (Sep, 2019 - version 0.94 + some upstream patches from git tree + some fixes for it to work on Python 3). On that time, I remember I checked that at least some of the bugs were due to big tables. There are several tables at the Linux Kernel that require several pages to be displayed. 

[errors.txt](https://github.com/rst2pdf/rst2pdf/files/5677549/errors.txt)

**Comment by mchehab** [comment]
FYI, there is an issue that it could be related to Python 3.9: it complains that "self.warn()" doesn't exist. I just added a quick hack here. See enclosed. Didn't test if this is backward-compatible with older versions.

In any case, this is unrelated to the reported issue.

[fix_warnings.txt](https://github.com/rst2pdf/rst2pdf/files/5677649/fix_warnings.txt)

**Comment by mchehab** [comment]
Btw, if you want to try compiling just one of the books, I just fixed a bug at the rst2pdf target and folded with the original one at this commit:

https://git.linuxtv.org/mchehab/experimental.git/commit/?h=rst2pdf-v3&id=b73c93c6946ab324443608fac62333b7e327a7e4

after that, you can simply request to build a single book:

```
$ make SPHINXDIRS=devicetree rst2pdf
```

On such case, the file will be written into a different directory (Documentation/output/<book>/pdf)
.

The devicetree seems to be the smallest one that it fails to build. It has only 11 files:

```
$ find Documentation/devicetree/ -name '*.rst'
Documentation/devicetree/overlay-notes.rst
Documentation/devicetree/index.rst
Documentation/devicetree/writing-schema.rst
Documentation/devicetree/changesets.rst
Documentation/devicetree/dynamic-resolution-notes.rst
Documentation/devicetree/usage-model.rst
Documentation/devicetree/of_unittest.rst
Documentation/devicetree/bindings/index.rst
Documentation/devicetree/bindings/writing-bindings.rst
Documentation/devicetree/bindings/submitting-patches.rst
Documentation/devicetree/bindings/ABI.rst
```

**Comment by stephenfin** [comment]
FWIW, the [OpenStack nova](https://opendev.org/openstack/nova) docs also fail with issues. I've spent some time looking into it but not enough. I guess this is a combination of missing support for some docutils features along with outright bugs. I'll investigate if I get a chance.

**Comment by nfraprado** [comment]
@mchehab I managed to fix the devicetree doc by moving the footnote to the end of the file, as shown in the following patch:

```
From b2fa7940c1a9b57c921f4a8508ac7cd256af746f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?N=C3=ADcolas=20F=2E=20R=2E=20A=2E=20Prado?=
 <nfraprado@protonmail.com>
Date: Fri, 25 Dec 2020 18:16:24 -0300
Subject: [PATCH] docs: dt: usage_model.rst: Move footnote to end of file
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Move the footnote to the end of the file in order to solve the following
error when building with rst2pdf:

[ERROR] pdfbuilder.py:159 Failed to build doc
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/rst2pdf/pdfbuilder.py", line 156, in write
    docwriter.write(doctree, destination)
  File "/usr/lib/python3.9/site-packages/docutils/writers/__init__.py", line 78, in write
    self.translate()
  File "/usr/lib/python3.9/site-packages/rst2pdf/pdfbuilder.py", line 587, in translate
    self.document.walkabout(visitor)
  File "/usr/lib/python3.9/site-packages/docutils/nodes.py", line 214, in walkabout
    if child.walkabout(visitor):
  File "/usr/lib/python3.9/site-packages/docutils/nodes.py", line 214, in walkabout
    if child.walkabout(visitor):
  File "/usr/lib/python3.9/site-packages/docutils/nodes.py", line 214, in walkabout
    if child.walkabout(visitor):
  [Previous line repeated 2 more times]
  File "/usr/lib/python3.9/site-packages/docutils/nodes.py", line 206, in walkabout
    visitor.dispatch_visit(self)
  File "/usr/lib/python3.9/site-packages/docutils/nodes.py", line 1995, in dispatch_visit
    return method(node)
  File "/usr/lib/python3.9/site-packages/rst2pdf/pdfbuilder.py", line 787, in visit_footnote
    fnr = self.footnotedict[id]
KeyError: 'usage-model_id3'

Signed-off-by: Nícolas F. R. A. Prado <nfraprado@protonmail.com>
---
 Documentation/devicetree/usage-model.rst | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/Documentation/devicetree/usage-model.rst b/Documentation/devicetree/usage-model.rst
index e1b42dc63f01..5574b05c6652 100644
--- a/Documentation/devicetree/usage-model.rst
+++ b/Documentation/devicetree/usage-model.rst
@@ -12,8 +12,6 @@ This article describes how Linux uses the device tree.  An overview of
 the device tree data format can be found on the device tree usage page
 at devicetree.org\ [1]_.
 
-.. [1] https://elinux.org/Device_Tree_Usage
-
 The "Open Firmware Device Tree", or simply Device Tree (DT), is a data
 structure and language for describing hardware.  More specifically, it
 is a description of hardware that is readable by an operating system
@@ -418,3 +416,5 @@ device creation model a little bit, but the solution turns out not to
 be too invasive.  If a node is compatible with "arm,amba-primecell", then
 of_platform_populate() will register it as an amba_device instead of a
 platform_device.
+
+.. [1] https://elinux.org/Device_Tree_Usage
-- 
2.29.2
```

There are two references to that footnote in the document, and moving the footnote declaration to anywhere after the second reference solves the issue. But I'm not sure if the original footnote position is allowed by docutils and therefore it should be fixed in rst2pdf instead, but at least that's a workaround.

I'll see if I can do anything about the other docs failing next.

**Comment by nfraprado** [comment]
Also, it's worth saying that all docs aren't actually failing with the same error. The `[ERROR] pdfbuilder.py:159 Failed to build doc` at the top is a catch-all saying that the document build failed, but the real error is actually shown at the bottom of the traceback. In the devicetree doc, for example, it was a footnote error:

```
File "/usr/lib/python3.9/site-packages/rst2pdf/pdfbuilder.py", line 787, in visit_footnote
    fnr = self.footnotedict[id]
KeyError: 'usage-model_id3'
```

while the one one originally posted by Mauro is something about a missing link:

```
  File "/devel/v4l/docs/sphinx_2.4.4/lib64/python3.9/site-packages/reportlab/pdfbase/pdfdoc.py", line 1778, in format
    if f is None: raise ValueError("format not resolved, probably missing URL scheme or undefined destination target for '%s'" % self.name)
ValueError: format not resolved, probably missing URL scheme or undefined destination target for 'sleeping-things'
```

**Comment by nfraprado** [comment]
In the case of the Kernel's RCU documentation, apart from having to install the `svglib` module and actual mistakes in the doc's markup that should be fixed on the kernel (and that weren't catched by the usual sphinx html and latex toolchains), there were errors with two images that were too big:

```
[ERROR] pdfbuilder.py:159 Failed to build doc
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/rst2pdf/pdfbuilder.py", line 156, in write
    docwriter.write(doctree, destination)
  File "/usr/lib/python3.9/site-packages/docutils/writers/__init__.py", line 78, in write
    self.translate()
  File "/usr/lib/python3.9/site-packages/rst2pdf/pdfbuilder.py", line 672, in translate
    createpdf.RstToPdf(
  File "/usr/lib/python3.9/site-packages/rst2pdf/createpdf.py", line 659, in createPdf
    pdfdoc.multiBuild(elements)
  File "/usr/lib/python3.9/site-packages/reportlab/platypus/doctemplate.py", line 1167, in multiBuild
    self.build(tempStory, **buildKwds)
  File "/usr/lib/python3.9/site-packages/reportlab/platypus/doctemplate.py", line 1080, in build
    self.handle_flowable(flowables)
  File "/usr/lib/python3.9/site-packages/rst2pdf/createpdf.py", line 813, in handle_flowable
    self.handle_frameEnd()
  File "/usr/lib/python3.9/site-packages/reportlab/platypus/doctemplate.py", line 726, in handle_frameEnd
    self.handle_pageEnd()
  File "/usr/lib/python3.9/site-packages/reportlab/platypus/doctemplate.py", line 668, in handle_pageEnd
    raise LayoutError(ident)
reportlab.platypus.doctemplate.LayoutError: More than 10 pages generated without content - halting layout.  Likely that a flowable is too large for any frame.
```

I noticed that rst2pdf already handles images that are too wide:

```
[WARNING] image.py:415 Image /home/nfraprado/cur/git/linux/Documentation/RCU/Design/Expedited-Grace-Periods/ExpSchedFlow.svg is too wide for the frame, rescaling
```

and on closer inspection I noticed that the images that cause the above error are rather tall, so it is my understanding that rst2pdf is able to work with images too wide, but not too tall, through rescaling. Wouldn't it be possible for rst2pdf to rescale an image by height once a full page was skipped? (Instead of just trying to fit it for 10 pages and giving up).

The two problematic images can be seen in the following places on the online kernel documentation:

https://www.kernel.org/doc/html/latest/RCU/Design/Expedited-Grace-Periods/Expedited-Grace-Periods.html#rcu-sched-expedited-grace-periods

https://www.kernel.org/doc/html/latest/RCU/Design/Memory-Ordering/Tree-RCU-Memory-Ordering.html#putting-it-all-together

**Comment by nfraprado** [comment]
Another recurrent issue with these documents is table size overflowing, but in this case, it is certainly a rst2pdf or reportlab issue.

I can reproduce it with the following minimal file:

```
PDF Documents From Your Text Editor
###################################

=  =======================================
0  Register as a non-blocking blk-mq drive
1  Register as a blocking blk-mq driver de
=  =======================================
```

Trying to run rst2pdf with the previous file, using the default stylesheet, causes the following error:

```
Traceback (most recent call last):
  File "/usr/bin/rst2pdf", line 8, in <module>
    sys.exit(main())
  File "/usr/lib/python3.9/site-packages/rst2pdf/createpdf.py", line 1683, in main
    RstToPdf(
  File "/usr/lib/python3.9/site-packages/rst2pdf/createpdf.py", line 659, in createPdf
    pdfdoc.multiBuild(elements)
  File "/usr/lib/python3.9/site-packages/reportlab/platypus/doctemplate.py", line 1167, in multiBuild
    self.build(tempStory, **buildKwds)
  File "/usr/lib/python3.9/site-packages/reportlab/platypus/doctemplate.py", line 1080, in build
    self.handle_flowable(flowables)
  File "/usr/lib/python3.9/site-packages/rst2pdf/createpdf.py", line 813, in handle_flowable
    self.handle_frameEnd()
  File "/usr/lib/python3.9/site-packages/reportlab/platypus/doctemplate.py", line 726, in handle_frameEnd
    self.handle_pageEnd()
  File "/usr/lib/python3.9/site-packages/reportlab/platypus/doctemplate.py", line 668, in handle_pageEnd
    raise LayoutError(ident)
reportlab.platypus.doctemplate.LayoutError: More than 10 pages generated without content - halting layout.  Likely that a flowable is too large for any frame.
```

Removing a single character from the table, it now becomes small enough to fit the page and the pdf is generated. But seeing the result pdf, it is clear that the table should have fitted with that extra character:

![2020-12-25_22:38:22](https://user-images.githubusercontent.com/43095212/103144070-f243c880-4701-11eb-822e-aed02568e08c.png)

**Comment by mchehab** [comment]
> @mchehab I managed to fix the devicetree doc by moving the footnote to the end of the file
> ...
> There are two references to that footnote in the document, and moving the footnote declaration to anywhere after the second reference solves the issue. But I'm not sure if the original footnote position is allowed by docutils and therefore it should be fixed in rst2pdf instead, but at least that's a workaround.

This is compliance problem with `rst2pdf` that it is not following ReST specs. Docutils spec explicitly says:

```
Footnotes may occur anywhere in the document, not only at the end.
Where and how they appear in the processed output depends on the processing system.
```
https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#footnotes

So, it is OK if rst2pdf internally places all of them at the end of the document, but the source is allowed to define footnotes anywhere.

There are lots of Kernel documents with footnotes. On almost all, the footnotes are just below the text which uses it, as this makes easier for people reading the file to see the footnotes. So, the actual fix should happen at rst2pdf (or at the libraries it may be relying on).

**Comment by lornajane** [comment]
I agree that the footnote workaround is only a workaround, and also that we'd be very happy to have a fix for this if anyone could advise us!

**Comment by nfraprado** [comment]
I submitted a fix for the footnote issue at #965.

**Comment by nfraprado** [comment]
Regarding the issue with images too tall, actually rst2pdf does support resizing tall images, and it does work fine. Using those same problematic images in a new document under sphinx, worked without a problem.

The issue is in the interaction between kernel's `kernel-figure` custom sphinx directive and rst2pdf, as changing that directive to a normal `image` in the kernel documentation makes it work.

**Comment by nfraprado** [comment]
Actually the issue isn't with `kernel-figure`, but `figure` in general, and I opened an issue for that at #966.

I think if #965 gets merged, and #964 and #966 get fixed, all Linux Kernel doc errors will be solved. Here's hoping.

**Comment by lornajane** [comment]
This is such a great summary of the moving parts on this issue! #965 is merged, let's see what else we can fix while we are at it :)

**Comment by lornajane** [comment]
The fix for #964 is in PR #979 (via duplicate issue #820, are you all keeping up with all this?), open but not merged.

**Comment by lornajane** [comment]
With the improvements to the internal linking merged now from #979 (with thanks to @nfraprado!) , we are left with #966 relating to tall images and and #967 relating to small tables that are still outstanding for the batch of issues reported here.

**Comment by lornajane** [comment]
Update: A fix for #967 has merged into master. Table width guessing is difficult to always get right but it shouldn't cause the document generation to fail so hopefully this should help, let us know?

**Comment by nfraprado** [comment]
I finally got around to testing if the table fix from #967 worked for the kernel docs and indeed it did! The document for the `block/` subsystem which would previously fail, now is successfully generated. I attached the resulting pdf here in case anyone is curious:  

[block.pdf](https://github.com/rst2pdf/rst2pdf/files/6080144/block.pdf)

[And here's the web version for reference](https://www.kernel.org/doc/html/latest/block/index.html).

And indeed I don't see any issues with the tables (there are some fluctuations to the right margin, but it's also like that on the web, so must be something with the .rst file or Sphinx itself). It seems rst2pdf did the job quite well! 

Now, I think it's only #966 left to go, which was breaking the docs of a lot of subsystems (`RCU`, `networking`, `GPU`, `userspace-api` and a few others)
